File: services/source-ingestion-service/src/markdown_normalizer.py
from __future__ import annotations

import re


def normalize_markdown(text: str) -> str:
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    text = text.replace("\ufeff", "")
    lines = text.split("\n")
    cleaned = []
    for line in lines:
        cleaned.append(line.rstrip())
    text = "\n".join(cleaned)
    text = re.sub(r"\n{3,}", "\n\n", text)
    text = re.sub(r"^#{1,6}(?=\S)", lambda m: m.group(0) + " ", text, flags=re.MULTILINE)
    text = text.rstrip("\n") + "\n"
    return text

File: services/source-ingestion-service/ts-src/markdownNormalizer.ts
const HEADING_PATTERN = /^#{1,6}(?=\S)/gm;
const NEWLINE_COLLAPSE_PATTERN = /\n{3,}/g;

export function normalizeMarkdown(text: string): string {
  text = text.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  text = text.replace(/\uFEFF/g, "");

  const lines = text.split("\n");
  const cleaned = lines.map((line) => line.replace(/[ \t]+$/, ""));
  text = cleaned.join("\n");

  text = text.replace(NEWLINE_COLLAPSE_PATTERN, "\n\n");
  text = text.replace(HEADING_PATTERN, (match) => match + " ");
  text = text.replace(/\n*$/, "") + "\n";

  return text;
}
